### PR TITLE
Allow painting of Apple Pay button in Mac GPUP

### PR DESCRIPTION
**Pull request by nmoucht**
#### 6b4a6f81a9aefaff2eb854ac7283c814a53a6c3b
<pre>
Allow painting of Apple Pay button in Mac GPUP
<a href="https://bugs.webkit.org/show_bug.cgi?id=247916">https://bugs.webkit.org/show_bug.cgi?id=247916</a>
&lt;rdar://102277273&gt;

Reviewed by Simon Fraser.

Allow painting of apple pay button without platform context
as it is not used by paintApplePayButton.

* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::canPaintWithoutPlatformContext):
(WebCore::RenderTheme::paint):

Canonical link: <a href="https://commits.webkit.org/256713@main">https://commits.webkit.org/256713@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/a358e1564404a271cdd0bfc1fded29d4d7bf95fb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96618 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5875 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29694 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106142 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100602 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6070 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34610 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88991 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102854 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102293 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83219 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31498 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88247 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40341 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/38004 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21142 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4656 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4273 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1086 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40419 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/rendering/RenderTheme.cpp [hunk 1]
```diff
@@ -462,14 +480,7 @@ bool RenderTheme::paint(const RenderBox& box, ControlStates& controlStates, cons
...
 
-    // Temporarily move this call above the canPaint check to allow
-    // this to work in the GPU process
-#if ENABLE(ATTACHMENT_ELEMENT)
-    if (part == AttachmentPart || part == BorderlessAttachmentPart)
-        return paintAttachment(box, paintInfo, integralSnappedRect);
-#endif
-
-    if (UNLIKELY(!canPaint(paintInfo, box.settings())))
+    if (!canPaintWithoutPlatformContext(part) && UNLIKELY(!canPaint(paintInfo, box.settings())))
```
Comment (smfr) [inline Source/WebCore/rendering/RenderTheme.cpp]: This is getting hard to follow. What about passing the ControlPart to `canPaint()`?

Comment (nmoucht) [reply Source/WebCore/rendering/RenderTheme.cpp]: Fixed.

File: Source/WebCore/rendering/RenderTheme.h [hunk 2]
```diff
@@ -252,7 +252,7 @@ class RenderTheme {
 #endif
 
 protected:
-    virtual bool canPaint(const PaintInfo&, const Settings&) const { return true; }
+    virtual bool canPaint(const PaintInfo&, const Settings&, const ControlPart) const { return true; }
```
Comment (smfr) [inline Source/WebCore/rendering/RenderTheme.h]: No need for `const` if you pass by value.

**Review by smfr — APPROVED** [review]

File: Source/WebCore/rendering/RenderTheme.h (hunk 2 above)
Comment (nmoucht) [reply Source/WebCore/rendering/RenderTheme.h]: Fixed.
